hazelcast/proxy/reliable_topic.py

```diff
@@ -1,12 +1,331 @@
+    def on_message(self, item):
+        """
+        Invoked when a message is received for the added reliable topic.
+
+        :param: message the message that is received for the added reliable topic
+        """
         raise NotImplementedError
 
+    def retrieve_initial_sequence(self):
+        """
+        Retrieves the initial sequence from which this ReliableMessageListener
+        should start.
+
+        Return -1 if there is no initial sequence and you want to start
+        from the next published message.
+
+        If you intend to create a durable subscriber so you continue from where
+        you stopped the previous time, load the previous sequence and add 1.
+        If you don't add one, then you will be receiving the same message twice.
+
+        :return: (int), the initial sequence
+        """
+        return -1
+
+    def store_sequence(self, sequence):
+        """"
+        Informs the ReliableMessageListener that it should store the sequence.
+        This method is called before the message is processed. Can be used to
+        make a durable subscription.
+
+        :param: (int) ``sequence`` the sequence
+        """
+        pass
+
+    def is_loss_tolerant(self):
+        """
+        Checks if this ReliableMessageListener is able to deal with message loss.
+        Even though the reliable topic promises to be reliable, it can be that a
+        MessageListener is too slow. Eventually the message won't be available
+        anymore.
+
+        If the ReliableMessageListener is not loss tolerant and the topic detects
+        that there are missing messages, it will terminate the
+        ReliableMessageListener.
+
+        :return: (bool) ``True`` if the ReliableMessageListener is tolerant towards losing messages.
+        """
+        return False
+
+    def is_terminal(self):
+        """
+        Checks if the ReliableMessageListener should be terminated based on an
+        exception thrown while calling on_message.
+
+        :return: (bool) ``True` if the ReliableMessageListener should terminate itself, ``False`` if it should keep on running.
+        """
+        raise False
+
+
+class _MessageListener(object):
+    def __init__(self, uuid, proxy, to_object, listener):
+        self._id = uuid
+        self._proxy = proxy
+        self._to_object = to_object
+        self._listener = listener
+        self._cancelled_lock = threading.Lock()
+        self._cancelled = False
+        self._sequence = 0
+        self._q = queue.Queue()
+
+    def start(self):
+        tail_seq = self._proxy.ringbuffer.tail_sequence()
+        initial_seq = self._listener.retrieve_initial_sequence()
+        if initial_seq == -1:
+            initial_seq = tail_seq.result() + 1
+        self._sequence = initial_seq
+        self._proxy.client.reactor.add_timer(0, self._next)
+
+    def _handle_illegal_argument_error(self):
+        head_seq = self._proxy.ringbuffer.head_sequence().result()
```

Comment:
So, if we are at this line, we are on the reactor thread. If you call `.result()` on reactor thread, you will have a deadlock (because the reactor thread itself is the one sets the results of futures). Therefore, there is a check on the future objects that forces you to use `add_done_callback` on scenarios like this. I believe with the current test suite, we are not hitting this line.